File: javascript/src/iterator/iterator.test.js
import assert from "node:assert/strict";
import { describe, it } from "node:test";
import { ConcreteAggregate } from "./concrete-aggregate.js";

function abc() {
  const aggregate = new ConcreteAggregate();
  for (const item of ["a", "b", "c"]) {
    aggregate.add(item);
  }
  return aggregate;
}

describe("Iterator", () => {
  it("traverses in order and stops at the end", () => {
    const iter = abc().createIterator();
    assert.equal(iter.hasNext(), true);
    assert.deepEqual([iter.next(), iter.next(), iter.next()], ["a", "b", "c"]);
    assert.equal(iter.hasNext(), false);
    assert.throws(() => iter.next(), /no more elements/);
  });

  it("iterators over one aggregate are independent", () => {
    const aggregate = abc();
    const first = aggregate.createIterator();
    const second = aggregate.createIterator();
    first.next();
    first.next();
    assert.equal(first.next(), "c");
    assert.equal(second.next(), "a");
  });

  it("an empty aggregate has nothing to visit", () => {
    const empty = new ConcreteAggregate();
    assert.equal(empty.count(), 0);
    assert.equal(empty.createIterator().hasNext(), false);
  });
});

File: java/src/test/java/com/penapereira/patterns/iterator/IteratorTest.java
package com.penapereira.patterns.iterator;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.NoSuchElementException;
import org.junit.jupiter.api.Test;

class IteratorTest {

    private static ConcreteAggregate abc() {
        ConcreteAggregate aggregate = new ConcreteAggregate();
        aggregate.add("a");
        aggregate.add("b");
        aggregate.add("c");
        return aggregate;
    }

    @Test
    void traversesInOrderAndStopsAtTheEnd() {
        Iterator it = abc().createIterator();
        assertTrue(it.hasNext());
        assertEquals("a", it.next());
        assertEquals("b", it.next());
        assertEquals("c", it.next());
        assertFalse(it.hasNext());
        assertThrows(NoSuchElementException.class, it::next);
    }

    @Test
    void iteratorsOverOneAggregateAreIndependent() {
        Aggregate aggregate = abc();
        Iterator first = aggregate.createIterator();
        Iterator second = aggregate.createIterator();
        first.next();
        first.next();
        assertEquals("c", first.next());
        assertEquals("a", second.next());
    }

    @Test
    void anEmptyAggregateHasNothingToVisit() {
        ConcreteAggregate empty = new ConcreteAggregate();
        assertEquals(0, empty.count());
        assertFalse(empty.createIterator().hasNext());
    }
}
